openmc/cmfd.py

```diff
@@ -0,0 +1,3072 @@
+        self._allocate_cmfd()
+
+        # Compute and store array indices used to build cross section
+        # arrays
+        self._precompute_array_indices()
+
+        # Compute and store row and column indices used to build CMFD
+        # matrices
+        self._precompute_matrix_indices()
+
+        # Initialize all variables used for linear solver in C++
+        self._initialize_linsolver()
+
+        # Initialize simulation
+        openmc.lib.simulation_init()
+
+        # Set cmfd_run variable to True through C API
+        openmc.lib.settings.cmfd_run = True
+
+    def next_batch(self):
+        """ Run next batch for CMFDRun.
+
+        Returns
+        -------
+        int
+            Status after running a batch (0=normal, 1=reached maximum number of
+            batches, 2=tally triggers reached)
+
+        """
+        # Initialize CMFD batch
+        self._cmfd_init_batch()
+
+        # Run next batch
+        status = openmc.lib.next_batch()
+
+        # Perform CMFD calculation if on
+        if self._cmfd_on:
+            self._execute_cmfd()
+
+            # Write CMFD output if CMFD on for current batch
+            if openmc.lib.master():
+                self._write_cmfd_output()
+
+        # Write CMFD data to statepoint
+        if openmc.lib.is_statepoint_batch():
+            self.statepoint_write()
+        return status
+
+    def finalize(self):
+        """ Finalize simulation by calling
+        :func:`openmc.lib.simulation_finalize` and print out CMFD timing
+        information.
+
+        """
+        # Finalize simuation
+        openmc.lib.simulation_finalize()
+
+        # Print out CMFD timing statistics
+        self._write_cmfd_timing_stats()
+
+    def statepoint_write(self, filename=None):
+        """Write all simulation parameters to statepoint
+
+        Parameters
+        ----------
+        filename : str
+            Filename of statepoint
+
+        """
+        if filename is None:
+            batch_str_len = len(str(openmc.lib.settings.batches))
+            batch_str = str(openmc.lib.current_batch()).zfill(batch_str_len)
+            filename = 'statepoint.{}.h5'.format(batch_str)
+
+        # Call C API statepoint_write to save source distribution with CMFD
+        # feedback
+        openmc.lib.statepoint_write(filename=filename)
+
+        # Append CMFD data to statepoint file using h5py
+        self._write_cmfd_statepoint(filename)
+
+    def _write_cmfd_statepoint(self, filename):
+        """Append all CNFD simulation parameters to existing statepoint
+
+        Parameters
+        ----------
+        filename : str
+            Filename of statepoint
+
+        """
+        if openmc.lib.master():
+            with h5py.File(filename, 'a') as f:
+                if 'cmfd' not in f:
+                    if openmc.lib.settings.verbosity >= 5:
+                        print(' Writing CMFD data to {}...'.format(filename))
+                        sys.stdout.flush()
```

Comment:
Question: Any reason you're explicitly flushing `stdout`, here and on line 905?  Does it have to do with the potential for program failure between here and line 905?